src/platforms/wayland/displayclient.cpp

```diff
@@ -0,0 +1,860 @@
+        bound_outputs.clear();
+    }
+    registry.reset();
+    wl_display_roundtrip(display);
+
+    eglDestroyContext(egldisplay, eglctx);
+    eglTerminate(egldisplay);
+}
+
+void mgw::DisplayClient::new_global(
+    void* data,
+    struct wl_registry* registry,
+    uint32_t id,
+    char const* interface,
+    uint32_t version)
+{
+    (void)version;
+    DisplayClient* self = static_cast<decltype(self)>(data);
+
+    if (strcmp(interface, "wl_compositor") == 0)
+    {
+        self->compositor =
+            static_cast<decltype(self->compositor)>(wl_registry_bind(registry, id, &wl_compositor_interface, std::min(version, 3u)));
+    }
+    else if (strcmp(interface, "wl_shm") == 0)
+    {
+        self->shm = static_cast<decltype(self->shm)>(wl_registry_bind(registry, id, &wl_shm_interface, std::min(version, 1u)));
+        // Normally we'd add a listener to pick up the supported formats here
+        // As luck would have it, I know that argb8888 is the only format we support :)
+        // {arg} TODO needs fixing
+        add_shm_listener(self, self->shm);
+    }
+    else if (strcmp(interface, "wl_seat") == 0)
+    {
+        if (version < 5) self->fake_pointer_frame = true;
+        self->seat = static_cast<decltype(self->seat)>(wl_registry_bind(registry, id, &wl_seat_interface, std::min(version, 6u)));
+        add_seat_listener(self, self->seat);
+    }
+    else if (strcmp(interface, "wl_output") == 0)
+    {
+        auto output =
+            static_cast<wl_output*>(wl_registry_bind(registry, id, &wl_output_interface, std::min(version, 2u)));
+        std::lock_guard<decltype(self->outputs_mutex)> lock{self->outputs_mutex};
+        self->bound_outputs.insert(
+            std::make_pair(
+                id,
+                std::make_unique<Output>(
+                    output,
+                    self,
+                    [self](Output const& output) { self->on_new_output(&output); },
+                    [self](Output const& output) { self->on_output_changed(&output); })));
+    }
+    else if (strcmp(interface, "wl_shell") == 0)
+    {
+        self->shell = static_cast<decltype(self->shell)>(wl_registry_bind(registry, id, &wl_shell_interface, std::min(version, 1u)));
+    }
+}
+
+void mgw::DisplayClient::remove_global(
+    void* data,
+    struct wl_registry* /*registry*/,
+    uint32_t id)
+{
+    DisplayClient* self = static_cast<decltype(self)>(data);
+
+    std::lock_guard<decltype(self->outputs_mutex)> lock{self->outputs_mutex};
+    auto const output = self->bound_outputs.find(id);
+    if (output != self->bound_outputs.end())
+    {
+        self->on_output_gone(output->second.get());
+        self->bound_outputs.erase(output);
+    }
+    // TODO: We should probably also delete any other globals we've bound to that disappear.
+}
+
+void mgw::DisplayClient::keyboard_keymap(wl_keyboard* /*keyboard*/, uint32_t /*format*/, int32_t fd, uint32_t size)
```

Comment:
Perhaps we should check to make sure `format` is `WL_KEYBOARD_KEYMAP_FORMAT_XKB_V1` (if that is what we are assuming)